educator/src/Dscp/DB/SQL/Util/Singleton.hs

```diff
@@ -0,0 +1,84 @@
+{-# LANGUAGE PartialTypeSignatures #-}
+{-# OPTIONS_GHC -Wno-partial-type-signatures #-}
+
+module Dscp.DB.SQL.Util.Singleton
+     ( SingletonT (..)
+     , getSingleton_
+     , setSingleton_
+     , loadSingleton_
+     ) where
+
+import Prelude hiding (_1, _2)
+
+import Data.Typeable (typeRep)
+import Database.Beam.Schema (PrimaryKey, TableEntity)
+import qualified Database.Beam.Schema as Beam
+import qualified Text.Show
+
+import Dscp.DB.SQL.Functions
+import Dscp.DB.SQL.Util.Common
+import Dscp.Util
+
+-- | Stores a single value.
+data SingletonT a (f :: * -> *) = SingletonUnsafe
+    { singletonId   :: Beam.C f Text
+      -- ^ Primary key, Beam cannot live without it.
+      -- Also it allows to easily extend the table to keep multiple elements.
+    , singletonItem :: Beam.C f a
+      -- ^ The stored item itself.
+    } deriving (Generic)
```

Comment:
Do I understand correctly that it is done in order to support JSON migrations for certificates? Not quite sure how exactly it would be used if the serialisation version is not included into JSON (using `Versioned`, for example`), and if we do store certificates in `Versioned` form, then I don't quite understand why do we need a global singleton value of currently supported version in a database (not in the code).

Reply:
Yes, this is for migrations. AFAIU we have to choices:
1. Attach a version to each certificate JSON, and migrate lazily on `GET` requests (otherwise, all versions would be always the same - the latest).
2. Keep the global version of all certificate JSONs, and migrate upon the node launch.

If we want to ever support certificate search then the first option becomes tricky: for instance, in a new version we add a new field and a search on this field, how will we perform this search if the field does not yet exist in old versions of certificate JSON?
Thus 2nd looks like the better option to me.

Reply:
@Martoon-00 Okay, so that means that we also need to have some constant version in the code, which represents the certificate version used by the code and should be matched with the value in DB on startup. Also, the migration logic itself should be provided somehow. When do you suggest to introduce all this?